lib/react.dart

```diff
@@ -86,32 +96,59 @@ abstract class Component {
 
   /// Force a call to [render] by calling [setState], which effectively "redraws" the `Component`.
   ///
+  /// Optionally accepts a callback that gets called after the component updates.
+  ///
   /// [A.k.a "forceUpdate"](https://facebook.github.io/react/docs/component-api.html#forceupdate)
-  void redraw() {
-    setState({});
+  void redraw([callback()]) {
+    setState({}, callback);
   }
 
   /// Set [_nextState] to provided [newState] value and force a re-render.
   ///
+  /// Optionally accepts a callback that gets called after the component updates.
+  ///
+  /// Also allows [newState] to be used as a transactional `setState` callback.
+  ///
   /// See: <https://facebook.github.io/react/docs/component-api.html#setstate>
-  void setState(Map newState) {
-    if (newState != null) {
+  void setState(dynamic newState, [callback()]) {
+    if (newState is Map) {
       _nextState.addAll(newState);
+    } else if (newState is _TransactionalSetStateCallback) {
+      _addTransactionalSetStateCallback(newState);
+    } else if (newState != null) {
+      throw new ArgumentError('setState expects its first parameter to either be a Map or a Function that accepts two parameters.');
     }
 
+    if (callback != null) _addSetStateCallback(callback);
+
     _jsRedraw();
   }
 
   /// Set [_nextState] to provided [newState] value and force a re-render.
   ///
+  /// Optionally accepts a callback that gets called after the component updates.
+  ///
   /// See: <https://facebook.github.io/react/docs/component-api.html#replacestate>
-  void replaceState(Map newState) {
+  void replaceState(Map newState, [callback()]) {
     Map nextState = newState == null ? {} : new Map.from(newState);
     _nextState = nextState;
+    if (callback != null) _addSetStateCallback(callback);
 
     _jsRedraw();
   }
 
+  void _addTransactionalSetStateCallback(Map callback(Map prevState, Map currentProps)) {
+    if (_transactionalSetStateCallbacks == null) _transactionalSetStateCallbacks = [];
+
+    _transactionalSetStateCallbacks.add(callback);
+  }
+
+  void _addSetStateCallback(callback()) {
+    if (_setStateCallbacks == null) _setStateCallbacks = [];
```

Comment:
Why are these null-checks necessary? I don't see `_setStateCallbacks` or `_transactionalSetStateCallbacks` ever being set to null.